Infer target version based on project metadata

stinodego opened the pull request:
<!-- Hello! Thanks for submitting a PR. To help make things go a bit more
     smoothly we would appreciate that you go through this template. -->

### Description

<!-- Good things to put here include: reasoning for the change (please link
     any relevant issues!), any noteworthy (or hacky) choices to be aware of,
     or what the problem resolved here looked like ... we won't mind a ranty
     story :) -->

Fixes #3124

The logic implemented is as follows:
* If Black's `target_version` is specified in the `pyproject.toml`, we ignore the project metadata.
* If it is not specified, we try to read and parse the project metadata to determine a valid list of `TargetVersion`. If we fail for whatever reason, we continue without a `target_version`.
* The list of target versions should include all major Python versions allowed by the `requires-python` specifier.
  * `3.8.5` becomes `["py38"]`
  * `>3.6,<3.11` becomes `["py37", "py38", "py39", "py310"]`
  * `<3.7` becomes `["py33", "py34", "py35", "py36"]` (Python 3.3 is the minimal supported version for Black)

_(for more examples, check the test cases)_

Changes:
* Updated `parse_pyproject_toml`
  * When reading the TOML file, we check if `target_version` is specified in Black's config. If not, we try to infer it from the project metadata.
  * This seemed like the logical place - since we are reading the file anyway, we might aswell extract the project metadata if we need it.
* Added basic functionality for reading project metadata.
  * Only the PyPA standard is now supported: `project` -> `requires-python`.
  * We could add support for Poetry or other tools later.
* Added parsing of the version / specifier based on `packaging`'s `Version` and `SpecifierSet`
  * There is some black magic (pun intended) going on in the `strip_specifier_set` function that I wrote. Unfortunately, working with these kinds of specifiers is nasty when it comes to edge cases. My implementation satisfies all specifier sets I could come up with, but there are probably some corner cases that I did not cover...
* Added `packaging>=20.0` as a dependency. This is the minimum version that includes the functionality needed for this implementation.
* Updated the help text for the CLI command.

### Checklist - did you ...

<!-- If any of the following items aren't relevant for your contribution
     please still tick them so we know you've gone through the checklist.

    All user-facing changes should get an entry. Otherwise, signal to us
    this should get the magical label to silence the CHANGELOG entry check.
    Tests are required for bugfixes and new features. Documentation changes
    are necessary for formatting and most enhancement changes. -->

- [x] Add a CHANGELOG entry if necessary?
- [x] Add / update tests if necessary?
- [x] Add new / update outdated documentation? _--> Updated the CLI argument help text_

<!-- Just as a reminder, everyone in all psf/black spaces including PRs
     must follow the PSF Code of Conduct (link below).

     Finally, once again thanks for your time and effort. If you have any
     feedback in regards to your experience contributing here, please
     let us know!

     Helpful links:

      PSF COC: https://www.python.org/psf/conduct/
      Contributing docs: https://black.readthedocs.io/en/latest/contributing/index.html
      Chat on Python Discord: https://discord.gg/RtVdv86PrH -->

JelleZijlstra commented on src/black/files.py:
```diff
@@ -103,14 +106,91 @@ def find_pyproject_toml(path_search_start: Tuple[str, ...]) -> Optional[str]:
+
+    If parsing fails, will raise a packaging.version.InvalidVersion error.
+    """
+    version = Version(requires_python)
```
Why doesn't this need to be in a try-except?

JelleZijlstra replied:
Oh I suppose it throws InvalidVersion in that case. That raises the question of when this raises `InvalidVersion` and when it returns None, if both of those represent an invalid version.

stinodego replied:
In this case, `InvalidVersion` represents an invalid version value entered for the `requires-python` field. We return None if the version is valid, but cannot be mapped to one of Black's TargetVersions. I updated the docstrings to clarify.

I created a separate function `parse_req_python` which handles the `Invalid` exceptions; I wanted to keep each function short to match the style of the other code in the module.

ichard26 commented:
> Add new / update outdated documentation? --> There isn't good documentation on --target-version, in my opinion. I could add this, but that could be considered out of scope for this PR. Let me know what you prefer!

Drive-by comment, for the time being you _could_ document this in the help for `--target-version` in the Click CLI setup decorators applied to `black.main`, but until I get #2839 landed, there are not many places you can document it. Getting that PR landed is on my priority to-do list, but I don't know whether I'll get it in before this PR.

felix-hilden commented on src/black/files.py:
```diff
@@ -103,14 +106,93 @@ def find_pyproject_toml(path_search_start: Tuple[str, ...]) -> Optional[str]:
+    If the field is not present or cannot be parsed, returns None.
+    """
+    project_metadata = pyproject_toml.get("project", {})
+    requires_python: Optional[str] = project_metadata.get("requires-python", None)
+    return requires_python
```
```suggestion
    return project_metadata.get("requires-python", None)
```

stinodego replied:
Unfortunately, mypy doesn't like your suggestion:

```
error: Returning Any from function declared to return "Optional[str]"  [no-any-return]
```

That's why I explicitly cast it to an optional string.

felix-hilden commented on tests/test_black.py:
```diff
@@ -1352,6 +1352,35 @@ def test_parse_pyproject_toml(self) -> None:
+            (">=3.10", TargetVersion.PY310),
+            (">3.6,<3.10", TargetVersion.PY37),
+            ("==3.8.*", TargetVersion.PY38),
+            # (">=3.8.6", TargetVersion.PY38),  # Doesn't work yet
```
Can we ignore/drop the patch version? Guess not at least easily, but a hack could be to construct a `SpecifierSet` using version strings with the patch sliced out, e.g. `v.split(".")[:2]` or something horrible.

stinodego replied:
I added a function with some black magic that does exactly that (but there were a bunch of special cases I had to handle). Now it covers all the cases I could think of!

felix-hilden reviewed (commented):
Thanks for the PR! Some small comments below.

stinodego commented:
Thanks for the comments so far! I came up with the missing logic for handling some of the special cases. The result isn't super pretty, but I believe it works. It covers all the test cases I could think of.

I set the PR to 'ready to review'. Happy to address any concerns you may have!

felix-hilden commented on tests/test_black.py:
```diff
@@ -1352,6 +1352,39 @@ def test_parse_pyproject_toml(self) -> None:
         self.assertEqual(config["exclude"], r"\.pyi?$")
         self.assertEqual(config["include"], r"\.py?$")
 
+    def test_parse_pyproject_toml_project_metadata(self) -> None:
```
This would be a good place for pytest parameterise instead of a loop!

stinodego replied:
That would be great, but unfortunately, `parametrize` doesn't work within this `BlackTestCase` class. It's a different testing framework.

Not sure why the tests are set up like this - there are different test files that do use normal pytest functions and parametrize - but for now I have to resort to this loop 😞

felix-hilden replied:
Oh that's right. Well it's fine this way as well.

felix-hilden commented on tests/test_black.py:
```diff
@@ -1352,6 +1352,39 @@ def test_parse_pyproject_toml(self) -> None:
+            config = black.parse_pyproject_toml(str(test_toml_file))
+            self.assertEqual(config.get("target_version"), expected)
+
+    def test_infer_target_version(self) -> None:
```
Likewise here, parameterise!

stinodego replied:
See my other comment. Unfortunately, not possible in this testing framework 😕

stinodego commented on tests/test_black.py:
```diff
@@ -1352,6 +1352,39 @@ def test_parse_pyproject_toml(self) -> None:
+            ("3.11.0rc1", TargetVersion.PY311),
+            (">=3.10", TargetVersion.PY310),
+            (">3.6,<3.10", TargetVersion.PY37),
+            ("==3.8.*", TargetVersion.PY38),
```
Technically, you could specify `==3.*`. That would allow any Python 3 version. It would result in `TargetVersion.PY33`.

stinodego replied:
Added!

stinodego commented:
@ichard26 Would you be willing to take another look at this?

We can also conclude that this kind of complexity is not what we want right now, for the small benefit it provides. Then we can close the PR. But would be nice to have some kind of decision on this!

ichard26 commented:
Hi sorry, I know it's been a long while since I've last said anything. I promise I haven't forgotten this PR, I'll try my best to review this PR in the coming days! On initial thought, I think this will be an useful QOL feature that the extra complexity is worth it.

stinodego commented:
I saw I missed a mypy linting error. I rebased the branch and added a commit to fix the issue.

stinodego commented:
Had to rebase due to #3233 . Should be good to go!

ichard26 commented:
I've been a really bad maintainer, but I promise I've started reviewing this. It overall looks great, I just have a few code style remarks and I want to play around with it locally before approving it. Thank you so much for being patient (way beyond what's reasonable!).

stinodego commented:
> I've been a really bad maintainer, but I promise I've started reviewing this. It overall looks great, I just have a few code style remarks and I want to play around with it locally before approving it. Thank you so much for being patient (way beyond what's reasonable!).

Don't worry about it! Life gets in the way. We'll just pick it back up whenever you're ready 👍

ichard26 commented on tests/test_black.py:
```diff
@@ -1352,6 +1352,40 @@ def test_parse_pyproject_toml(self) -> None:
+            ("==3.*", TargetVersion.PY33),
+            (">=3.8.6", TargetVersion.PY38),
+            ("> 3.7.4, != 3.8.8", TargetVersion.PY37),
+            (">3.7,!=3.8", TargetVersion.PY39),
```
```suggestion
            (">3.7,!=3.8", TargetVersion.PY39),
            ("!=3.3,!=3.5", TargetVersion.PY35),
```

stinodego replied:
Your test case doesn't work (returns TargetVersion.PY34). I think you meant `"!=3.3,!=3.4"`. I added this test case in a new commit.

ichard26 commented on src/black/files.py:
```diff
@@ -108,14 +111,123 @@ def find_pyproject_toml(path_search_start: Tuple[str, ...]) -> Optional[str]:
+    If parsing fails, will raise a packaging.specifiers.InvalidSpecifier error.
+    If the parsed specifier cannot be mapped to a valid TargetVersion, returns None.
+    """
+    specifier_set = strip_specifier_set(SpecifierSet(requires_python))
```
Black will blow up with `Error: Could not open file '/home/ichard26/programming/oss/black/pyproject.toml': Error reading configuration file: Invalid version: 'a'` if you pass a specifier set that has an invalid version that's sanitized (eg. `>=3.7,!=a`).

Given the defensive style of the rest of the logic, it's probably best to return None if `InvalidVersion` is raised here. The other option is to only ignore invalid specifiers by catching `InvalidVersion` in `strip_specifier_set`'s loop.

stinodego replied:
Good catch - let's indeed return None in this case. I will figure out how to do this nicely and add a commit for this.

stinodego replied:
This is now fixed!

ichard26 commented on src/black/files.py:
```diff
@@ -108,14 +111,123 @@ def find_pyproject_toml(path_search_start: Tuple[str, ...]) -> Optional[str]:
+    return requires_python
+
+
+def parse_req_python(requires_python: str) -> Optional[TargetVersion]:
```
I merged this and the other function into `infer_target_version` locally. Lemme know what you think:

```python
def infer_target_version(pyproject_toml: Dict[str, Any]) -> Optional[TargetVersion]:
    """Infer Black's target version from the project metadata in pyproject.toml.

    Currently only supports the PyPA standard format (PEP 621):
    https://packaging.python.org/en/latest/specifications/declaring-project-metadata/#requires-python

    If target version cannot be inferred, returns None.
    """
    project_metadata = pyproject_toml.get("project", {})
    requires_python: Optional[str] = project_metadata.get("requires-python", None)
    if requires_python is not None:
        try:
            return parse_req_python_version(requires_python)
        except InvalidVersion:
            pass
        try:
            return parse_req_python_specifier(requires_python)
        except InvalidSpecifier:
            pass

    return None
```

stinodego replied:
Much better this way. Added a commit with this change.

ichard26 commented on src/black/__init__.py:
```diff
@@ -218,8 +218,10 @@ def validate_regex(
     help=(
-        "Python versions that should be supported by Black's output. [default: per-file"
-        " auto-detection]"
+        "Minimum Python version that should be supported by Black's output. By default,"
+        " Black will try to infer this from the project metadata in pyproject.toml. If"
+        " this does not yield conclusive results, Black will use per-file"
+        " auto-detection."
```
Good call to document the per-file auto-detection fallback!

ichard26 commented on CHANGES.md:
```diff
@@ -27,6 +27,9 @@
 
 <!-- Changes to how Black can be configured -->
 
+- Black now tries to infer its `--target-version` from the project metadata specified in
+  pyproject.toml (#3219)
```
```suggestion
- Black now tries to infer its `--target-version` from the project metadata specified in
  `pyproject.toml` (#3219)
```

stinodego replied:
Added this in the latest commit.

ichard26 reviewed (changes requested):
Fantastic work overall! Your tests are great and your code was easy to follow. As mentioned earlier, I do have a few stylistic critiques and suggestions to make the logic more robust.

Also two general comments:

- Should we emit a message telling the user what we inferred the target version to be as part of `--verbose`? 
- By inferring a default for `--target-version`, Black will never do per-file auto-detection if `project.requires-python` exists and is valid. This is certainly better than the status quo for the lower bound, but it can cause some parsing trouble if the file in question is actually using more modern syntax than specified.[^1]

  Say for example we have `requires-python = ">=3.7"`. If I format a file that's using 3.10's new match statement, Black will fail with a `Cannot parse: ...` error because [Black decides what grammars it'll try from the target version](https://github.com/psf/black/blob/f16333e78ba77ab29ab0b75e10891e6c65c6da7e/src/black/parsing.py#L49-L79). While this keeps things consistent, I don't think it makes for a good user experience. The main two ways to fix this are either to infer a list of all supported target versions or [actually implement this suggestion to make `--target-version` a lower-bound by default](https://github.com/psf/black/issues/751#issuecomment-1168882896). The latter is the long-term fix, but I'm not sure when that will be implemented so the former might be necessary as a temporary fix.[^2]

Thanks again for waiting so patiently <3

[^1]: Also Black would never introduce modern syntax that's not permissible under the project's inferred target version even if the file in question is using more modern syntax -- which is probably fine given explicitly providing `--target-version` works the same way.
[^2]: Unless we decide this is an acceptable failure case and the workaround of explicitly setting `--target-version` is okay -- which is fine by me

stinodego commented:
Thanks for your comments! I just got back from vacation and will take a look at this tomorrow.

stinodego commented on src/black/files.py:
```diff
@@ -108,14 +111,123 @@ def find_pyproject_toml(path_search_start: Tuple[str, ...]) -> Optional[str]:
+    If parsing fails, will raise a packaging.specifiers.InvalidSpecifier error.
+    If the parsed specifier cannot be mapped to a valid TargetVersion, returns None.
+    """
+    specifier_set = strip_specifier_set(SpecifierSet(requires_python))
+
+    if not specifier_set:
+        return None
```
Solved this in the latest commit.

stinodego commented on src/black/files.py:
```diff
@@ -108,14 +111,123 @@ def find_pyproject_toml(path_search_start: Tuple[str, ...]) -> Optional[str]:
+    return config
+
+
+def infer_target_version(pyproject_toml: Dict[str, Any]) -> Optional[TargetVersion]:
+    """Infer Black's target version from the project metadata in pyproject.toml.
+
+    If target version cannot be inferred, returns None.
+    """
+    requires_python = get_req_python(pyproject_toml)
+    if requires_python is not None:
+        return parse_req_python(requires_python)
+    return None
+
+
+def get_req_python(pyproject_toml: Dict[str, Any]) -> Optional[str]:
+    """Get the required Python version from the project metadata.
+
+    Currently only supports the PyPA standard format:
+    https://packaging.python.org/en/latest/specifications/declaring-project-metadata/#requires-python
+
+    If the field is not present or cannot be parsed, returns None.
+    """
+    project_metadata = pyproject_toml.get("project", {})
+    requires_python: Optional[str] = project_metadata.get("requires-python", None)
+    return requires_python
```
I fully agree. I will follow your suggestion below.

stinodego commented on src/black/files.py:
```diff
@@ -108,14 +111,123 @@ def find_pyproject_toml(path_search_start: Tuple[str, ...]) -> Optional[str]:
+        return None
+
+    target_version_map = {f"3.{v.value}": v for v in TargetVersion}
+    compatible_versions = specifier_set.filter(target_version_map)
```
It's possible, but it's a minor performance hit. I think it's worth it, though it's slightly inelegant. I'll implement this in a new commit.

stinodego commented:
> Should we emit a message telling the user what we inferred the target version to be as part of `--verbose`?

Great suggestion, I will figure out how to do this and add a commit for it.

> * By inferring a default for `--target-version`, Black will never do per-file auto-detection if `project.requires-python` exists and is valid. This is certainly better than the status quo for the lower bound, but it can cause some parsing trouble if the file in question is actually using more modern syntax than specified.  [...] The main two ways to fix this are either to infer a list of all supported target versions or [actually implement this suggestion to make `--target-version` a lower-bound by default](https://github.com/psf/black/issues/751#issuecomment-1168882896). The latter is the long-term fix, but I'm not sure when that will be implemented so the former might be necessary as a temporary fix.

I hadn't considered this; thanks for explaining this. I think it shouldn't be hard to rewrite my logic to generate the full list of target versions. I'll figure this out and add a commit for this.

ichard26 reviewed (commented):
Almost there! The problem where Black blows up with invalid python-requires is still there.

`--verbose` still doesn't say what target version were inferred, but I'm happy to defer this as a future improvement since the per-file autodetection isn't logged anywhere either. A quick note about the changelog otherwise. 

Thank you so much for your patience again!

stinodego commented:
Hey @ichard26 , sorry I have been sidetracked with some other things - I will complete this sometime next week.

stinodego commented on CHANGES.md:
```diff
@@ -18,6 +18,9 @@
 
 <!-- Changes to how Black can be configured -->
 
+- Black now tries to infer its `--target-version` from the project metadata specified in
```
Added this!

stinodego commented on tests/test_black.py:
```diff
@@ -1512,6 +1512,73 @@ def test_parse_pyproject_toml(self) -> None:
+            config = black.parse_pyproject_toml(str(test_toml_file))
+            self.assertEqual(config.get("target_version"), expected)
+
+    @pytest.mark.filterwarnings("ignore::DeprecationWarning")
```
We create invalid specifiers to test the code, which causes `packaging` to throw a DeprecationWarning. So in order to test that these invalid specifiers return `None` as expected, we need to ignore these warnings.

stinodego replied:
No longer required with packaging >= 22.0

stinodego commented:
@ichard26 I addressed your comments, I think this handles all cases properly now. Happy to address any other feedback.

I would also be interested in doing some work as a follow-up to make Black use a single target version, as discussed previously.

felix-hilden reviewed (approved):
Thank you for addressing the comments, and sorry for the late re-review! This is looking good to me 👌

stinodego commented:
Thanks! I'll do a rebase to resolve a conflict in the changelog.

stinodego commented:
`packaging` version 22.0 removed some deprecated behaviour - that makes this implementation a bit cleaner. Had to do minor adjustments to get everything to work nice.

Note that lint now fails because of a new lint added by `flake8-bugbear`. I'll rerun this check once it's fixed.

ichard26 commented:
Oh gosh I've ignored this for two months. If I don't review this by this weekend, everyone has my consent to ping me about this PR. I imagine it'd be a simple approval, but yeah I've been busy with life.

stinodego commented:
No worries, I've been busy working on [Polars](https://github.com/pola-rs/polars).

Lint passes now after rebase. I don't know how to make diff-shades happy, though.

neutrinoceros commented:
@ichard26 any chance this could make it into the upcoming release ? :-)

ichard26 reviewed (approved):
Let's ship this! :ship:  There some tweaks that will probably have to be made, but I'm sick of letting this rot anymore. I'll deal with them in follow up PRs.

It'll be really nice to have this in 23.1.0. Thank you so, so, so much @stinodego for baring with me. 

>  @ ichard26 I addressed your comments, I think this handles all cases properly now. Happy to address any other feedback.

I would also be interested in doing some work as a follow-up to make Black use a single target version, as discussed previously.

That would be wonderful, although as you have noticed, it can get complicated quickly. You've been warned :)

ichard26 commented:
Wait, it seems like the attrs failure on diff-shades is genuine. I thought it was just a fluke, but it turns out attrs doesn't specify Black's target-version in `pyproject.toml` so evidently this PR is changing behaviour (in a bad way). I've marked this as draft as I investigate.

ichard26 commented:
Alright, so the reason why this is breaking attrs is that a non-empty `mode.target_versions` causes `black.parsing.get_grammars()` to return a single grammar. I haven't checked, but it's probably this one:

https://github.com/psf/black/blob/f16333e78ba77ab29ab0b75e10891e6c65c6da7e/src/black/parsing.py#L66-L69 

I miscontextualized how `get_grammars()` works. I thought as long as the 3.10 grammar might be supported, it'll be added to the list to try... that's not the case! Instead it checks for strict compatibility (using `black.mode.supports_feature()`) so if there's a target-version that doesn't support the 3.10 grammar, it won't be added.

Attrs is formatting fine right now because if `mode.target_versions` is empty, all grammars get picked:

https://github.com/psf/black/blob/f16333e78ba77ab29ab0b75e10891e6c65c6da7e/src/black/parsing.py#L50-L59

So technically this PR doesn't introduce any more regressions, but it just exposes this pitfall... I'm not entirely sure what the best way of fixing this is. My suggestion would be to add the 3.10 grammar if **any** of the target versions support `Feature.PATTERN_MATCHING`, not **all**.

ichard26 commented:
```diff
[attrs - https://github.com/python-attrs/attrs.git]
╰─> [revision 9d23ae9536beb3cd30241f05123190ec1ecc20c1](https://github.com/python-attrs/attrs/tree/9d23ae9536beb3cd30241f05123190ec1ecc20c1)
--- a/attrs:src/attr/validators.pyi
+++ b/attrs:src/attr/validators.pyi
@@ -80,7 +80,7 @@
 def min_len(length: int) -> _ValidatorType[_T]: ...
 def not_(
     validator: _ValidatorType[_T],
     *,
     msg: Optional[str] = None,
-    exc_types: Union[Type[Exception], Iterable[Type[Exception]]] = ...
+    exc_types: Union[Type[Exception], Iterable[Type[Exception]]] = ...,
 ) -> _ValidatorType[_T]: ...
 ```
 I don't have time to investigate, but presumably this is due to better inference of the target version. Not ideal, but as long as we land this in 23.1.0 where we're changing the stable style anyway, it should be _fine_.

ichard26 commented:
Thanks again @stinodego. It's been great working with you :black_heart:

stinodego commented:
Thanks for putting the finishing touches on this, @ichard26 ! Glad to have this part of the latest release.

If any issues pop up, do not hesitate to tag me.
